// atom-languageclient/build/test/helpers.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createFakeEditor = exports.createSpyConnection = void 0;
const sinon = require("sinon");
const atom_1 = require("atom");
function createSpyConnection() {
    return {
        listen: sinon.spy(),
        onClose: sinon.spy(),
        onError: sinon.spy(),
        onDispose: sinon.spy(),
        onUnhandledNotification: sinon.spy(),
        onUnhandledProgress: sinon.spy(),
        onRequest: sinon.spy(),
        onNotification: sinon.spy(),
        onProgress: sinon.spy(),
        dispose: sinon.spy(),
        sendRequest: sinon.spy(),
        sendNotification: sinon.spy(),
        sendProgress: sinon.spy(),
        trace: sinon.spy(),
        inspect: sinon.spy(),
    };
}
exports.createSpyConnection = createSpyConnection;
function createFakeEditor(path) {
    const editor = new atom_1.TextEditor();
    sinon.stub(editor, 'getSelectedBufferRange');
    sinon.spy(editor, 'setTextInBufferRange');
    editor.setTabLength(4);
    editor.setSoftTabs(true);
    editor.getBuffer().setPath(path || '/a/b/c/d.js');
    return editor;
}
exports.createFakeEditor = createFakeEditor;
