// atom-languageclient/build/lib/main.js
"use strict";
/* eslint-disable @typescript-eslint/triple-slash-reference */
/// <reference path="../typings/atom/index.d.ts"/>
/// <reference path="../typings/atom-ide/index.d.ts"/>
/* eslint-enable @typescript-eslint/triple-slash-reference */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.LinterPushV2Adapter = exports.DownloadFile = exports.FilteredLogger = exports.ConsoleLogger = exports.Convert = exports.AutoLanguageClient = void 0;
const auto_languageclient_1 = require("./auto-languageclient");
exports.AutoLanguageClient = auto_languageclient_1.default;
const convert_1 = require("./convert");
exports.Convert = convert_1.default;
const logger_1 = require("./logger");
Object.defineProperty(exports, "ConsoleLogger", { enumerable: true, get: function () { return logger_1.ConsoleLogger; } });
Object.defineProperty(exports, "FilteredLogger", { enumerable: true, get: function () { return logger_1.FilteredLogger; } });
const download_file_1 = require("./download-file");
exports.DownloadFile = download_file_1.default;
const linter_push_v2_adapter_1 = require("./adapters/linter-push-v2-adapter");
exports.LinterPushV2Adapter = linter_push_v2_adapter_1.default;
__exportStar(require("./auto-languageclient"), exports);
