// atom-languageclient/build/test/adapters/signature-help-adapter.test.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const atom_1 = require("atom");
const signature_help_adapter_1 = require("../../lib/adapters/signature-help-adapter");
const helpers_1 = require("../helpers");
const chai_1 = require("chai");
const sinon = require("sinon");
describe('SignatureHelpAdapter', () => {
    describe('canAdapt', () => {
        it('checks for signatureHelpProvider', () => {
            chai_1.expect(signature_help_adapter_1.default.canAdapt({})).to.equal(false);
            chai_1.expect(signature_help_adapter_1.default.canAdapt({ signatureHelpProvider: {} })).to.equal(true);
        });
    });
    describe('can attach to a server', () => {
        it('subscribes to onPublishDiagnostics', () => __awaiter(void 0, void 0, void 0, function* () {
            const connection = helpers_1.createSpyConnection();
            connection.signatureHelp = sinon.stub().resolves({ signatures: [] });
            const adapter = new signature_help_adapter_1.default({
                connection,
                capabilities: {
                    signatureHelpProvider: {
                        triggerCharacters: ['(', ','],
                    },
                },
            }, ['source.js']);
            const spy = sinon.stub().returns(new atom_1.Disposable());
            adapter.attach(spy);
            chai_1.expect(spy.calledOnce).to.be.true;
            const provider = spy.firstCall.args[0];
            chai_1.expect(provider.priority).to.equal(1);
            chai_1.expect(provider.grammarScopes).to.deep.equal(['source.js']);
            chai_1.expect(provider.triggerCharacters).to.deep.equal(new Set(['(', ',']));
            chai_1.expect(typeof provider.getSignatureHelp).to.equal('function');
            const result = yield provider.getSignatureHelp(helpers_1.createFakeEditor('test.txt'), new atom_1.Point(0, 1));
            chai_1.expect(connection.signatureHelp.calledOnce).to.be.true;
            const params = connection.signatureHelp.firstCall.args[0];
            chai_1.expect(params).to.deep.equal({
                textDocument: { uri: 'file:///test.txt' },
                position: { line: 0, character: 1 },
            });
            chai_1.expect(result).to.deep.equal({ signatures: [] });
        }));
    });
});
