// vls/dist/utils/cancellationToken.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isVCancellationRequested = exports.VCancellationTokenSource = void 0;
const vscode_languageserver_1 = require("vscode-languageserver");
class VCancellationTokenSource extends vscode_languageserver_1.CancellationTokenSource {
    constructor(tsModule) {
        super();
        this.tsModule = tsModule;
    }
    get token() {
        const operationCancelException = this.tsModule.OperationCanceledException;
        const token = super.token;
        token.tsToken = {
            isCancellationRequested() {
                return token.isCancellationRequested;
            },
            throwIfCancellationRequested() {
                if (token.isCancellationRequested) {
                    throw new operationCancelException();
                }
            }
        };
        return token;
    }
}
exports.VCancellationTokenSource = VCancellationTokenSource;
function isVCancellationRequested(token) {
    return new Promise(resolve => {
        if (!token) {
            resolve(false);
        }
        else {
            setImmediate(() => resolve(token.isCancellationRequested));
        }
    });
}
exports.isVCancellationRequested = isVCancellationRequested;
